File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private static TreeNode construct(int [] nums, int left, int right){
        if (left > right) return null;
        int mid = left + (right - left) / 2;
        return  new TreeNode(nums[mid], 
                                                      construct(nums, left, mid - 1), 
                                                      construct(nums, mid + 1, right));
    }
    public TreeNode sortedArrayToBST(int[] nums) {
        return construct(nums, 0, nums.length - 1);
    }
}

File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def construct(left, right):
            nonlocal n
            if left > right:
                    return None
            mid = (left + right) // 2
            rt = TreeNode(nums[mid], construct(left, mid - 1), construct(mid + 1, right));
            return rt
        n = len(nums)
        return construct(0, n - 1)